src/testing/cmds/harness/test-runner/main.go

```diff
@@ -0,0 +1,182 @@
+package main
+
+import (
+	"bufio"
+	"context"
+	"flag"
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+	"strings"
+	"time"
+
+	"github.com/google/shlex"
+	"github.com/pachyderm/pachyderm/v2/src/internal/backoff"
+	"github.com/pachyderm/pachyderm/v2/src/internal/errors"
+	"github.com/pachyderm/pachyderm/v2/src/internal/log"
+	"github.com/pachyderm/pachyderm/v2/src/internal/pctx"
+	"go.uber.org/zap"
+	"golang.org/x/exp/slices"
+	"golang.org/x/sync/errgroup"
+)
+
+func main() {
+	log.InitPachctlLogger()
+	ctx := pctx.Background("test-runner")
+	tags := flag.String("tags", "", "Tags to run, for example k8s. Tests without this flag will not be selected.")
+	fileName := flag.String("file", "tests_to_run.csv", "csv file containing the list of test names to run generated by the collector.")
+	gotestsumArgsRaw := flag.String("gotestsum-args", "", "Additional arguments to pass to the gotestsum portion of the test command.")
+	gotestArgsRaw := flag.String("gotest-args", "", "Additional arguments to pass to the 'go test' portion of the test command.")
+	shard := flag.Int("shard", 0, "0 indexed current runner index that we are on.")
+	totalShards := flag.Int("total-shards", 1, "Total number of runners that we are sharding over.")
+	threadPool := flag.Int("threads", 1, "Number of tests to execute concurrently.")
+	flag.Parse()
+	gotestsumArgs, err := shlex.Split(*gotestsumArgsRaw)
+	if err != nil {
+		log.Error(ctx, "Error parsing gotestsumArgs", zap.Error(err))
+		os.Exit(1)
+	}
+	gotestArgs, err := shlex.Split(*gotestArgsRaw)
+	if err != nil {
+		log.Error(ctx, "Error parsing gotestArgs", zap.Error(err))
+		os.Exit(1)
+	}
+
+	err = run(ctx,
+		*tags,
+		*fileName,
+		gotestsumArgs,
+		gotestArgs,
+		*shard,
+		*totalShards,
+		*threadPool,
+	)
+	if err != nil {
+		log.Error(ctx, "Error running tests", zap.Error(err))
+		os.Exit(1)
+	}
+	os.Exit(0)
+}
+
+func run(ctx context.Context, tags string, fileName string, gotestsumArgs []string, gotestArgs []string, shard int, totalShards int, threadPool int) error {
+	tests, err := readTests(ctx, fileName)
+	if err != nil {
+		return errors.Wrapf(err, "reading file %v", fileName)
+	}
+	slices.Sort(tests) // sort so we shard them from a pre-determined order
+
+	// loop through by the number of shards so that each gets a roughly equal number on each
+	testsForShard := map[string][]string{}
+	for idx := shard; idx < len(tests); idx += totalShards {
+		val := strings.Split(tests[idx], ",")
+		pkg := val[0]
```

Comment:
I like to assert that `len(val} > 1` before indexing into strings.Split output.  That way, I can get an informative error message instead of a panic.